New page: aria-invalid (#10278)
mdn/content: files/en-us/web/accessibility/aria/attributes/aria-invalid/index.md

---
title: 'aria-invalid'
slug: Web/Accessibility/ARIA/Attributes/aria-invalid
tags: 
  - Accessibility
  - ARIA
  - ARIA attribute
  - ARIA property
  - aria-invalid
  - Reference
---

The `aria-invalid` state indicates the entered value does not conform to the format expected by the application. 

## Description

The `aria-invalid` attribute is used to indicate that the value entered into an input field is not in a format or a value the application will accept. This may include formats such as email addresses or telephone numbers. `aria-invalid` can also be used to indicate that a required field is empty.

> The `aria-invalid` attribute is not limited to elements that have an ARIA role assigned and can be used with any typical HTML form element.

Reviewer: ```suggestion
The `aria-invalid` attribute can be used with any typical HTML form element, and is not limited to elements that have an ARIA role assigned.
```

The attribute should be set with JavaScript as a result of a validation process. If a value is determined to be invalid or out-of-range, set `aria-invalid="true"` **and** tell the user there is an error. For better user experience, provide suggestions for how the error can be corrected. Do not set `aria-invalid="true"` on empty required elements until after the user attempts to submit the form. They may still be working on filling it out.

> **Note:** When `aria-invalid` is used in conjunction with the `aria-required` attribute, `aria-invalid` should not be set to true before the form is submitted - only in response to validation.

> There are currently four values: in addition to `true` and `false` we have `grammar` which can be used when a grammatical is detected and `spelling` for spelling errors.  If the attribute is not present, or its value is false, or its value is an empty string, the default value of false applies. Any other value is treated as if `true` were set. It is possible that more values will be added in the future.

Reviewer: ```suggestion
There are currently four values for `aria-invalid.` In addition to `true` and `false`, there is `grammar`, which can be used when a grammatical is detected, as well as `spelling`, for spelling errors. If the attribute is not present, or its value is `false,` or its value is an empty string, the default value of `false` applies. Any other value is treated as if `true` were set. 
```

### Native HTML validation

HTML has native form validation. When a user submits a form with a control containing errors, the first form control with an invalid value displays an error message, natively.

> If there is a {{htmlattrxref("required")}} attribute on a form control that isn't filled out, the form will not submit, and an error message appears reading "Please fill out this field" or something similar. 

Reviewer: ```suggestion
If there is a {{htmlattrxref("required")}} attribute on a form control that isn't filled out, the form will not submit, and an error message appears reading "Please fill out this field" or something similar. The messaging for native validation varies depending on the browser, and cannot be styled.
```

```html
<input type="number" step="2" min="0" max="100" required>
```

> If that form had a [number input type](/en-US/docs/Web/HTML/Element/input/number) and the user had entered a value above the maximum, below the minimum, or that doesn't match the step value, an error message would appear. For the input above, if the user had entered "3", the native error message would be similar to "Please enter a valid value. The two nearest valid values are 2 and 4."  

Reviewer: ```suggestion
If the user had entered a value in the preceding input example above the maximum, below the minimum, or that doesn't match the step value, an error message would appear. If the user had entered "3", the native error message would be similar to "Please enter a valid value."  
```

> The first rule of ARIA use is "if you can use a native feature with the semantics and behavior you require already built in, instead of repurposing an element and **adding** an ARIA role, state or property to make it accessible, then do so." If you use native form controls, the `aria-invalid` attribute isn't needed. In addition to native form validation, we get access to the {{CSSXRef(':invalid')}} pseudoclass and other [UI-Pseudoclasses](/en-US/docs/Learn/Forms/UI_pseudo-classes) for free. 
> 
> However, if for some reason you are creating your own form validation scripts, make sure to include `aria-invalid` on invalid form controls, along with styling (use the `[aria-invalid="true"]` attribute selector) and messaging (with [`aria-errormessage`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-errormessage)) to help users understand where the mistake is and how they can fix it.
> 
> 

Reviewer: ```suggestion
If you are creating your own form validation scripts, make sure to include `aria-invalid` on invalid form controls, along with styling (use the `[aria-invalid="true"]` attribute selector) and messaging (with [`aria-errormessage`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-errormessage)) to help users understand where the mistake is and how they can fix it.

```

Removing the first rule of ARIA bit here, as native is oftentimes not sufficient for intuitive form validation, which is bad from a cognitive accessibility standpoint.
